--- Day 15: Warehouse Woes ---
You appear back inside your own mini submarine! Each Historian drives their mini submarine in a different direction; maybe the Chief has his own submarine down here somewhere as well?

You look up to see a vast school of lanternfish swimming past you. On closer inspection, they seem quite anxious, so you drive your mini submarine over to see if you can help.

Because lanternfish populations grow rapidly, they need a lot of food, and that food needs to be stored somewhere. That's why these lanternfish have built elaborate warehouse complexes operated by robots!

These lanternfish seem so anxious because they have lost control of the robot that operates one of their most important warehouses! It is currently running amok, pushing around boxes in the warehouse with no regard for lanternfish logistics or lanternfish inventory management strategies.

Right now, none of the lanternfish are brave enough to swim up to an unpredictable robot so they could shut it off. However, if you could anticipate the robot's movements, maybe they could find a safe option.

The lanternfish already have a map of the warehouse and a list of movements the robot will attempt to make (your puzzle input). The problem is that the movements will sometimes fail as boxes are shifted around, making the actual movements of the robot difficult to predict.

For example:

##########
#..O..O.O#
#......O.#
#.OO..O.O#
#..O@..O.#
#O#..O...#
#O..O..O.#
#.OO.O.OO#
#....O...#
##########

<vv>^<v^>v>^vv^v>v<>v^v<v<^vv<<<^><<><>>v<vvv<>^v^>^<<<><<v<<<v^vv^v>^
vvv<<^>^v^^><<>>><>^<<><^vv^^<>vvv<>><^^v>^>vv<>v<<<<v<^v>^<^^>>>^<v<v
><>vv>v^v^<>><>>>><^^>vv>v<^^^>>v^v^<^^>v^^>v^<^v>v<>>v^v^<v>v^^<^^vv<
<<v<^>>^^^^>>>v^<>vvv^><v<<<>^^^vv^<vvv>^>v<^^^^v<>^>vvvv><>>v^<<^^^^^
^><^><>>><>^^<<^^v>>><^<v>^<vv>>v>>>^v><>^v><<<<v>>v<v<v>vvv>^<><<>^><
^>><>^v<><^vvv<^^<><v<<<<<><^v<<<><<<^^<v<^^^><^>>^<v^><<<^>>^v<v^v<v^
>^>>^v>vv>^<<^v<>><<><<v<<v><>v<^vv<<<>^^v^>^^>>><<^v>>v^v><^^>>^<>vv^
<><^^>^^^<><vvvvv^v<v<<>^v<v>v<<^><<><<><<<^^<<<^<<>><<><^^^>^^<>^>v<>
^^>vv<^v^v<vv>^<><v<^v>^^^>>>^^vvv^>vvv<>>>^<^>>>>>^<<^v>^vvv<>^<><<v>
v^^>>><<^^<>>^v^<v^vv<>v^<<>^<^v^v><^<<<><<^<v><v<>vv>>v><v^<vv<>v^<<^
As the robot (@) attempts to move, if there are any boxes (O) in the way, the robot will also attempt to push those boxes. However, if this action would cause the robot or a box to move into a wall (#), nothing moves instead, including the robot. The initial positions of these are shown on the map at the top of the document the lanternfish gave you.

The rest of the document describes the moves (^ for up, v for down, < for left, > for right) that the robot will attempt to make, in order. (The moves form a single giant sequence; they are broken into multiple lines just to make copy-pasting easier. Newlines within the move sequence should be ignored.)

Here is a smaller example to get started:

########
#..O.O.#
##@.O..#
#...O..#
#.#.O..#
#...O..#
#......#
########

<^^>>>vv<v>>v<<
Were the robot to attempt the given sequence of moves, it would push around the boxes as follows:

Initial state:
########
#..O.O.#
##@.O..#
#...O..#
#.#.O..#
#...O..#
#......#
########

Move <:
########
#..O.O.#
##@.O..#
#...O..#
#.#.O..#
#...O..#
#......#
########

Move ^:
########
#.@O.O.#
##..O..#
#...O..#
#.#.O..#
#...O..#
#......#
########

Move ^:
########
#.@O.O.#
##..O..#
#...O..#
#.#.O..#
#...O..#
#......#
########

Move >:
########
#..@OO.#
##..O..#
#...O..#
#.#.O..#
#...O..#
#......#
########

Move >:
########
#...@OO#
##..O..#
#...O..#
#.#.O..#
#...O..#
#......#
########

Move >:
########
#...@OO#
##..O..#
#...O..#
#.#.O..#
#...O..#
#......#
########

Move v:
########
#....OO#
##..@..#
#...O..#
#.#.O..#
#...O..#
#...O..#
########

Move v:
########
#....OO#
##..@..#
#...O..#
#.#.O..#
#...O..#
#...O..#
########

Move <:
########
#....OO#
##.@...#
#...O..#
#.#.O..#
#...O..#
#...O..#
########

Move v:
########
#....OO#
##.....#
#..@O..#
#.#.O..#
#...O..#
#...O..#
########

Move >:
########
#....OO#
##.....#
#...@O.#
#.#.O..#
#...O..#
#...O..#
########

Move >:
########
#....OO#
##.....#
#....@O#
#.#.O..#
#...O..#
#...O..#
########

Move v:
########
#....OO#
##.....#
#.....O#
#.#.O@.#
#...O..#
#...O..#
########

Move <:
########
#....OO#
##.....#
#.....O#
#.#O@..#
#...O..#
#...O..#
########

Move <:
########
#....OO#
##.....#
#.....O#
#.#O@..#
#...O..#
#...O..#
########
The larger example has many more moves; after the robot has finished those moves, the warehouse would look like this:

##########
#.O.O.OOO#
#........#
#OO......#
#OO@.....#
#O#.....O#
#O.....OO#
#O.....OO#
#OO....OO#
##########
The lanternfish use their own custom Goods Positioning System (GPS for short) to track the locations of the boxes. The GPS coordinate of a box is equal to 100 times its distance from the top edge of the map plus its distance from the left edge of the map. (This process does not stop at wall tiles; measure all the way to the edges of the map.)

So, the box shown below has a distance of 1 from the top edge of the map and 4 from the left edge of the map, resulting in a GPS coordinate of 100 * 1 + 4 = 104.

#######
#...O..
#......
The lanternfish would like to know the sum of all boxes' GPS coordinates after the robot finishes moving. In the larger example, the sum of all boxes' GPS coordinates is 10092. In the smaller example, the sum is 2028.

Predict the motion of the robot and boxes in the warehouse. After the robot is finished moving, what is the sum of all boxes' GPS coordinates?

In [69]:
"""
Data structure:
An array (list of lists) with one list per line in the grid and number entries for the fields occupied by a box
A similar array for walls
For each move we iteratively look forward to see if the space is empty, a wall or box (iterate if box)
if move we move robot position and we move boxes (just need to remove last box and add one new box in the empty space to move into)
"""

'\nData structure:\nAn array (list of lists) with one list per line in the grid and number entries for the fields occupied by a box\nA similar array for walls\nFor each move we iteratively look forward to see if the space is empty, a wall or box (iterate if box)\nif move we move robot position and we move boxes (just need to remove last box and add one new box in the empty space to move into)\n'

In [70]:
with open('input15.txt') as file:
    boxes = []
    walls = []
    robot_moves = []
    for x, line in enumerate(file):
        if line.startswith('#'):
            # grid:
            boxes.append([])
            walls.append([])
            for y, pos in enumerate(line):
                if pos == '#':
                    walls[-1].append(y)
                elif pos == 'O':
                    boxes[-1].append(y)
                elif pos == '@':
                    rob_pos = (x, y)

        else:
            for m in line:
                if m == '>' or m == '<' or m == 'v' or m == '^':
                    robot_moves.append(m)
grid_size = len(boxes)
len(boxes), len(walls[0]), robot_moves[:10], walls[1], rob_pos

(50,
 50,
 ['>', 'v', '>', '>', '>', '^', '>', '>', 'v', '>'],
 [0, 1, 9, 17, 22, 27, 29, 49],
 (24, 24))

In [71]:
def move(orig, dir):
    x, y = orig
    if dir == '<':
        y -= 1
    elif dir == '>':
        y += 1
    elif dir == '^':
        x -= 1
    elif dir == 'v':
        x += 1
    return (x, y)


def look_ahead(orig, dir):
    """returns pos of next empty space, None if no space (wall)"""
    x, y = move(orig, dir)
    if y in walls[x]:
        return None
    elif y in boxes[x]:
        return look_ahead((x,y), dir)
    else: # space
        return (x, y)
    
look_ahead((26,24), 'v')

(31, 24)

In [72]:
for m in robot_moves:
    next_space = look_ahead(rob_pos, m)
    if next_space:
        rob_pos = move(rob_pos, m)
        if rob_pos != next_space:
            boxes[rob_pos[0]].remove(rob_pos[1])
            boxes[next_space[0]].append(next_space[1])

In [73]:
def gps(pos):
    x, y = pos
    return 100 * x + y

gps_sum = 0
for x, line in enumerate(boxes):
    for y in line:
        gps_sum += gps((x, y))
gps_sum

1406628

Your puzzle answer was 1406628.

The first half of this puzzle is complete! It provides one gold star: *

--- Part Two ---
The lanternfish use your information to find a safe moment to swim in and turn off the malfunctioning robot! Just as they start preparing a festival in your honor, reports start coming in that a second warehouse's robot is also malfunctioning.

This warehouse's layout is surprisingly similar to the one you just helped. There is one key difference: everything except the robot is twice as wide! The robot's list of movements doesn't change.

To get the wider warehouse's map, start with your original map and, for each tile, make the following changes:

If the tile is #, the new map contains ## instead.
If the tile is O, the new map contains [] instead.
If the tile is ., the new map contains .. instead.
If the tile is @, the new map contains @. instead.
This will produce a new warehouse map which is twice as wide and with wide boxes that are represented by []. (The robot does not change size.)

The larger example from before would now look like this:

####################
##....[]....[]..[]##
##............[]..##
##..[][]....[]..[]##
##....[]@.....[]..##
##[]##....[]......##
##[]....[]....[]..##
##..[][]..[]..[][]##
##........[]......##
####################
Because boxes are now twice as wide but the robot is still the same size and speed, boxes can be aligned such that they directly push two other boxes at once. For example, consider this situation:

#######
#...#.#
#.....#
#..OO@#
#..O..#
#.....#
#######

<vv<<^^<<^^
After appropriately resizing this map, the robot would push around these boxes as follows:

Initial state:
##############
##......##..##
##..........##
##....[][]@.##
##....[]....##
##..........##
##############

Move <:
##############
##......##..##
##..........##
##...[][]@..##
##....[]....##
##..........##
##############

Move v:
##############
##......##..##
##..........##
##...[][]...##
##....[].@..##
##..........##
##############

Move v:
##############
##......##..##
##..........##
##...[][]...##
##....[]....##
##.......@..##
##############

Move <:
##############
##......##..##
##..........##
##...[][]...##
##....[]....##
##......@...##
##############

Move <:
##############
##......##..##
##..........##
##...[][]...##
##....[]....##
##.....@....##
##############

Move ^:
##############
##......##..##
##...[][]...##
##....[]....##
##.....@....##
##..........##
##############

Move ^:
##############
##......##..##
##...[][]...##
##....[]....##
##.....@....##
##..........##
##############

Move <:
##############
##......##..##
##...[][]...##
##....[]....##
##....@.....##
##..........##
##############

Move <:
##############
##......##..##
##...[][]...##
##....[]....##
##...@......##
##..........##
##############

Move ^:
##############
##......##..##
##...[][]...##
##...@[]....##
##..........##
##..........##
##############

Move ^:
##############
##...[].##..##
##...@.[]...##
##....[]....##
##..........##
##..........##
##############
This warehouse also uses GPS to locate the boxes. For these larger boxes, distances are measured from the edge of the map to the closest edge of the box in question. So, the box shown below has a distance of 1 from the top edge of the map and 5 from the left edge of the map, resulting in a GPS coordinate of 100 * 1 + 5 = 105.

##########
##...[]...
##........
In the scaled-up version of the larger example from above, after the robot has finished all of its moves, the warehouse would look like this:

####################
##[].......[].[][]##
##[]...........[].##
##[]........[][][]##
##[]......[]....[]##
##..##......[]....##
##..[]............##
##..@......[].[][]##
##......[][]..[]..##
####################
The sum of these boxes' GPS coordinates is 9021.

Predict the motion of the robot and boxes in this new, scaled-up warehouse. What is the sum of all boxes' final GPS coordinates?

In [74]:
with open('input15.txt') as file:
    boxes = []
    walls = []
    for x, line in enumerate(file):
        if line.startswith('#'):
            # grid:
            boxes.append(set())
            walls.append(set())
            for y, pos in enumerate(line):
                y *= 2
                if pos == '#':
                    walls[-1].update([y, y+1])
                elif pos == 'O':
                    boxes[-1].add(y) # the new double spaced boxes are stored with their starting index
                elif pos == '@':
                    rob_pos = (x, y)

In [75]:
"""
Our old efficient approach will not work here. We need to find the whole cascade of boxes affected. If a single one cannot move nothing moves. If we can move, everything moves at once.
"""

'\nOur old efficient approach will not work here. We need to find the whole cascade of boxes affected. If a single one cannot move nothing moves. If we can move, everything moves at once.\n'

In [ ]:
for dir in robot_moves:
    queue = [move(rob_pos, dir)]
    cascade = set()
    movable = True
    while queue:
        x, y = queue.pop()
        if y in walls[x]:
            movable = False
            break
        
        # Box check:
        b_y = None
        if y-1 in boxes[x]:
            b_y = y-1
        elif y in boxes[x]:
            b_y = y
        if b_y is not None and (x, b_y) not in cascade:
            new_box = (x, b_y)
            cascade.add(new_box)
            if dir == '>':
                queue.append((x, b_y + 2))
            else:
                queue.append(move(new_box, dir))
            if dir == '^' or dir == 'v':
                queue.append(move((x, b_y + 1), dir))

    if movable:
        rob_pos = move(rob_pos, dir)
        for x, y in cascade:
            boxes[x].remove(y)
        for x, y in cascade:
            if dir == '^':
                boxes[x - 1].add(y)
            elif dir == 'v':
                boxes[x+1].add(y)
            elif dir == '<':
                boxes[x].add(y-1)
            elif dir == '>':
                boxes[x].add(y+1)

In [77]:
gps_sum = 0
for x, line in enumerate(boxes):
    for y in line:
        gps_sum += gps((x, y))
gps_sum

1432781

That's the right answer! You are one gold star closer to finding the Chief Historian.